### ISSUE TITLE
Question on memory leaks with nested models

### ISSUE THREAD

**Issue by truemetal** [Mantle/Mantle#721]
I notice memory leaks when using nested models in swift (declared in parent model like this `dynamic var nested = NestedModel()`)

Any fix around that ?

**Comment by JanGorman** [comment]
Could you share some more of your code or create an example that replicates this? Swift users ARC too so there really shouldn't be any difference compared to Objective-C.

**Comment by truemetal** [comment]
Does not occur when declaring like this `dynamic var nested:NestedModel!`, though I'd still prefer having that default value. Attaching my example project, and leak screenshot from instruments. 

![screen shot 2016-06-28 at 16 51 56](https://cloud.githubusercontent.com/assets/2472186/16418194/87ff10d8-3d51-11e6-9d39-447c7d14b5f8.png)

[mantle leak test.zip](https://github.com/Mantle/Mantle/files/337170/mantle.leak.test.zip)

**Comment by JanGorman** [comment]
Interesting, definitely something fishy going on here. If you really insist on having a variable initialised for your `NestedModel` class, you can use a backing variable like this, that won't leak:

```swift
private var _nested: NestedModel!
var nested: NestedModel {
  get {
    if _nested == nil {
      _nested = NestedModel()
    }
    return _nested
  }
  set {
    _nested = newValue
  }
}
```
and change your mapping dictionary to use `"nested": "_nested"`. To be honest, I'd make this an optional though and unwrap whenever you need it.

That aside, in `MTLModel` if you remove the empty initialiser that just calls super and change `- (instancetype)initWithDictionary:(NSDictionary *)dictionary error:(NSError **)error {` to call `[super init]` instead of `[self init]` that also fixes the leak. Rather weird. Unfortunately I don't have more time today to look into this. If anyone wants to pick it up

**Comment by truemetal** [comment]
@JanGorman I'd agree with you, probably better unwrap the optional then have that extra wrapper property with backing var. 

The solution with changing `[self init]` to `[super init]` seems to perfectly resolve the issue; will you change that yourself or should I add a pull-request?

**Comment by truemetal** [comment]
@JanGorman thanks for looking into this

**Comment by JanGorman** [comment]
@truemetal Best to open a PR, reference this issue. Thanks!

**Comment by truemetal** [comment]
@JanGorman that was quite a simple one :) 
Looking forward an update on cocoapods

**Comment by truemetal** [comment]
ok, looks like that change fails one of the tests
that '1' is being set in init() of MTLTestModel which gets bypassed by the change 

![mtlmodelspec_m](https://cloud.githubusercontent.com/assets/2472186/16449597/16ef6e82-3e01-11e6-9d82-0d65660cab9e.png)

**Comment by dcaunt** [comment]
Lets discuss the PR/proposed fix in #72 so it's easy to follow.

I'm also keen to understand why this code creates a leak as described here and whether it represents a bug in Swift/Objective-C tooling.
